hypothesis-python/src/hypothesis/searchstrategy/strings.py

```diff
@@ -63,19 +61,35 @@ def __init__(self,
                 'combination of arguments: ' + ', '.join(
                     '%s=%r' % arg for arg in arguments if arg[1] is not None)
             )
-        self.intervals = IntervalSet(intervals)
-        if whitelist_characters:
-            self.whitelist_characters = set(whitelist_characters)
-        else:
-            self.whitelist_characters = set()
-        self.zero_point = self.intervals.index_above(ord('0'))
+        self.category_intervals = [IntervalSet(i) for i in intervals]
+
+        all_intervals = sorted(sum(intervals, ()), reverse=True)
+        self.all_intervals = IntervalSet(
+            charmap._union_intervals(all_intervals[:1], all_intervals[1:])
+        )
+        self.zero_points = {iset: iset.index_above(ord('0'))
+                            for iset in self.category_intervals}
+        self.zero_points[None] = self.all_intervals.index_above(ord('0'))
 
     def do_draw(self, data):
+        category = choice(data, self.category_intervals)
+        # Almost always - when generating characters - we use the category
+        # chosen above and draw a codepoint from that category.  Then, we
+        # write the integer that would draw this from *all* categories.
+        if biased_coin(data, p=0.001):
```

Comment:
If you change this arbitrary value to something with a power-of-2 denominator (e.g. `p=1/1024`), then `biased_coin` should switch to using a single fixed-size draw, which will theoretically make shrinking easier.